### PR TITLE
don't produce duplicate outputs in `HLTPFJetsMatchedToFilteredJetsProducer`

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

As reported by HLT-JME contacts in a recent TSG meeting, the plugin `HLTPFJetsMatchedToFilteredJetsProducer` can produce duplicates in its output product.

This was already the case in the previous version of this plugin, i.e. [`PFJetsMatchedToFilteredCaloJetsProducer`](https://github.com/cms-sw/cmssw/blob/CMSSW_12_2_X/HLTrigger/JetMET/src/PFJetsMatchedToFilteredCaloJetsProducer.cc), and it was intentionally not changed in #37057 (which was merely a technical migration, for renaming purposes).

I think `12_6_X` would be a good time to fix this. Since there is no physics validation done yet, I'm not sure this change should be backported to `12_4_X` and `12_5_X` (since there are dozens of instances of `HLTPFJetsMatchedToFilteredJetsProducer` running at HLT).

_Edit_ : changes are expected (see https://github.com/cms-sw/cmssw/pull/39179#issuecomment-1241663125).

 - This PR removes duplicates in certain collections of HLT jets.
 - DQM modules that use the HLT jets passing certain jet triggers thus will have generally less entries (less jets), or jet-multiplicity distributions shifted to lower values.
 - In principle, even trigger results could change (1 jet duplicated N times would lead to passing a N-jet trigger before this PR, but not after the PR).
 - Differences might be more visible for lower-pT triggers, while at higher-pT I would expect less duplicates (and also we will have less stats to judge).

#### PR validation:

Next to none.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

N/A

### CODE REVIEW

File: HLTrigger/JetMET/interface/HLTPFJetsMatchedToFilteredJetsProducer.h [hunk 1]
```diff
@@ -58,7 +58,7 @@ void HLTPFJetsMatchedToFilteredJetsProducer<TriggerJetsRefType>::fillDescription
     edm::ConfigurationDescriptions& descriptions) {
   edm::ParameterSetDescription desc;
   desc.add<edm::InputTag>("src", edm::InputTag("hltPFJets"));
-  desc.add<edm::InputTag>("triggerJetsFilter", edm::InputTag("hltSingleJet240Regional"));
+  desc.add<edm::InputTag>("triggerJetsFilter", edm::InputTag("hltCaloJetsFiltered"));
```
Comment (missirol) [inline HLTrigger/JetMET/interface/HLTPFJetsMatchedToFilteredJetsProducer.h:61]: This default value is unimportant for HLT's purposes (always overriden in the configuration). `hltCaloJetsFiltered` just seemed less confusing/arbitrary than `hltSingleJet240Regional`.

File: HLTrigger/JetMET/interface/HLTPFJetsMatchedToFilteredJetsProducer.h [hunk 2]
```diff
@@ -71,19 +71,17 @@ void HLTPFJetsMatchedToFilteredJetsProducer<TriggerJetsRefType>::produce(edm::St
...
 
   auto outPFJets = std::make_unique<reco::PFJetCollection>();
   outPFJets->reserve(recoCands.size());
 
-  for (auto const& iJetRef : triggerJetsRefVec) {
-    for (auto const& jRecoCand : recoCands) {
-      auto const dR2 = reco::deltaR2(jRecoCand.p4(), iJetRef->p4());
-      if (dR2 < maxDeltaR2_) {
+  for (auto const& jRecoCand : recoCands) {
+    for (auto const& iJetRef : triggerJetsRefVec) {
```
Comment (missirol) [inline HLTrigger/JetMET/interface/HLTPFJetsMatchedToFilteredJetsProducer.h:83]: This re-ordering is the fix to avoid duplicates.

**Comment by missirol** [comment]
please test

**Comment by missirol** [comment]
hold

Outputs of PR tests look as expected (a few less entries in plots of HLT-jets).

Since the discussion in [CMSHLT-2450](https://its.cern.ch/jira/browse/CMSHLT-2450) concerns this plugin, this PR stays on hold until that discussion converges.

**Comment by missirol** [comment]
unhold

The discussion in [CMSHLT-2450](https://its.cern.ch/jira/browse/CMSHLT-2450) became largely unrelated to this PR (removal of duplicates). Since the fix is simple, but not considered necessary for earlier cycles, I plan to integrate it in `master` without any backports.

**Comment by missirol** [comment]
+hlt

**Comment by perrotta** [comment]
@missirol are all the changes observed in the tests for HLT DQM compatible with this PR?
You don't mention them in the description...

**Comment by missirol** [comment]
@perrotta 

>  You don't mention them in the description...

Sorry, you are right. Changes are expected.

 - This PR removes duplicates in certain collections of HLT jets.
 - DQM modules that use the HLT jets passing certain jet triggers thus have generally less entries (less jets), or jet-multiplicity distributions shifted to lower values.
 - In principle, even trigger results could change (1 jet duplicated N times would lead to passing a N-jet trigger before this PR, but not after the PR), but I suspect this is rare enough, and we don't see differences in trigger results in PR tests.
 - In DQM, we see differences in lower-pT jet triggers (e.g. PFJet40), which I think is reasonable (at higher pT, I expect less duplicates, although arguably we will also have less stats at higher pT in PR tests to judge this).

I will add this to the PR description.

Imho, the fix to the logic of the plugin is not controversial and it should go in `master`, but since there are physics changes without proper physics studies, I would not backport it.

To be extra sure, I will ask a review by @cms-sw/jetmet-pog-l2 .

**Comment by missirol** [comment]
assign jetmet-pog

@cms-sw/jetmet-pog-l2 , this PR fixes the logic inside a JME HLT plugin, to avoid duplicate jets in output.

Please review, and sign it if you agree. This PR targets `12_6_0_pre2`, deadline Sep-13.

For more information, see https://github.com/cms-sw/cmssw/pull/39179#issuecomment-1241663125.

**Comment by missirol** [comment]
@cms-sw/jetmet-pog-l2

Could you please review this PR?

**Comment by miquork** [comment]
I've checked the comments in this PR thread and am happy with them, having made the original request to have the duplicates removed. Thanks for the fix! I'll sign as soon as I find the button for it. Still new to handling PRs for JME.

**Comment by missirol** [comment]
@miquork , thanks for having a look; to approve, I think you can write a comment here with the very first line being exactly "+jme-pog" or "+1".

FYI: @sparedes

**Comment by miquork** [comment]
+1

Useful fix

**Comment by perrotta** [comment]
+1

**Pull request merged**